app/src/main/java/de/fraunhofer/fokus/OpenMobileNetworkToolkit/Preferences/SharedPreferencesGrouper.java

```diff
@@ -74,21 +78,88 @@ public SharedPreferences getSharedPreference(SPType key){
         return sp;
     }
 
-    public void removeListener(SPType key){
+    /**
+     * Removes a particular {@code listener} for a particular {@code key}
+     * @param key the key holding the listener
+     * @param listener the listener to remove
+     */
+    public synchronized void removeListener(@NonNull SPType key, @NonNull SharedPreferences.OnSharedPreferenceChangeListener listener) {
         SharedPreferences sp = this.getSharedPreference(key);
         if(sp == null) {
             Log.e(TAG, "SharedPreferences not found for "+key);
             return;
         }
-        SharedPreferences.OnSharedPreferenceChangeListener listener = this.spMap.get(key);
-        if(listener == null) {
+        Set<SharedPreferences.OnSharedPreferenceChangeListener> listenerList = this.spMap.get(key);
+        if(listenerList == null || !listenerList.contains(listener)) {
```

Comment:
ah but please. we can't just blatantly apply the exact same logic to three different methods with different purposes.

this one that you're commenting on is removing a concrete KEY-VALUE pair.
the map has a set of values for each key, and this method merely removes the passed-in value from the set held under that key. that is why we're checking if the set even contains the value.

the method on 118-119 removes all values for a key, so we don't need to check whether the set contains anything because we're planning on deleting the whole set from the map.

The method on 144-145 is deleting a single listener from an unknown/any key, so it looks through the whole map, each key at a time, and checks their sets whether they contain the listener.

The method on 177-178 is for adding a new listener, and it has to do that by first:
- get the set from the map under the passed-in key
- if null, create a new set
- add the passed-in listener to the set
- save the new set in the map under teh passed-in key, overwriting the previous value


So please, learn from this comment: don't just look at text that looks similar, and learn not to blatantly apply suggestions to text that looks similar - look for similar contexts. In this particular case, the slightly broader context than the two lines you're focusing on is completely different for all three methods, as they have a different purpose.

Please, pay attention to the broader picture a bit more.